src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -84,43 +94,136 @@ private Task OnProjectFactoryCompletedAsync()
+        }
+
+        private async Task OnProjectChangedCoreAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e, RuleHandlerType handlerType)
+        {
             // TODO: https://github.com/dotnet/roslyn-project-system/issues/353
             await _commonServices.ThreadingService.SwitchToUIThread();
 
             using (_tasksService.LoadedProject())
             {
-                await HandleAsync(e, handlerType).ConfigureAwait(false);
+                await HandleAsync(e, handlerType, _aggregateProjectContext).ConfigureAwait(false);
+            }
+
+            // If "TargetFrameworks" property has changed, we need to refresh the project context and subscriptions.
+            string newTargetFrameworks;
+            if (HasTargetFrameworksChanged(e, out newTargetFrameworks))
+            {
+                await UpdateProjectContextAndSubscriptionsAsync(newTargetFrameworks).ConfigureAwait(false);
+            }
+        }
+
+        private async Task UpdateProjectContextAndSubscriptionsAsync(string targetFrameworks)
+        {
+            lock (_projectContextProvider)
+            {
+                if (string.Equals(targetFrameworks, _latestTargetFrameworks, StringComparison.OrdinalIgnoreCase))
+                {
+                    // We have already handled this targetFrameworks update.
+                    return;
+                }
+            }
+
+            var newProjectContext = await _projectContextProvider.UpdateProjectContextAsync(targetFrameworks, CancellationToken.None).ConfigureAwait(false);
+
+            AggregateWorkspaceProjectContext previousProjectContext;
+            lock (_projectContextProvider)
+            {
+                if (newProjectContext == _aggregateProjectContext)
+                {
+                    // Another thread has already completed the update.
+                    Requires.Range(string.Equals(targetFrameworks, _latestTargetFrameworks, StringComparison.OrdinalIgnoreCase), nameof(targetFrameworks));
+                    return;
+                }
+
+                previousProjectContext = _aggregateProjectContext;
+                _aggregateProjectContext = newProjectContext;
+                _latestTargetFrameworks = targetFrameworks;
+            }
+
+            await ResetSubscriptionsAsync().ConfigureAwait(false);
+
+            if (previousProjectContext != null)
+            {
+                foreach (var innerContext in previousProjectContext.InnerProjectContexts)
+                {
+                    foreach (var handler in Handlers)
+                    {
+                        await handler.Value.OnContextReleasedAsync(innerContext).ConfigureAwait(false);
+                    }
+                }
+            }
+        }
+
+        private async Task ResetSubscriptionsAsync()
+        {
+            DisposeAndClearSubscriptions();
+
+            using (_tasksService.LoadedProject())
```

Comment:
>LoadedProject [](start = 33, length = 13)

Is LoadedProject() guard valid here since there is some asynchronous work happening?